align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+
+            selected_icl_examples = [possible_icl_examples[i] for i in indices]
+        else:
+            raise ValueError(f'"{icl_strategy}" is not a valid incontext method. Please use "random" or '
+                                '"bert_similarity"')
+
+        return selected_icl_examples
+
+
+class ComparativeRegressionIncontextExampleGenerator(IncontextExampleGenerator):
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+    
+    def set_icl_datasets(self):
+        icl_datasets = {}
+        incontext_data = self.read_icl_dataset_files()
+        
+        # Add each target to icl_datasets
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            icl_datasets[sys_kdma_name] = []
+            kdma_incontext_data = incontext_data[sys_kdma_name]
+            
+            # Add each examples to icl_datasets
+            for example in kdma_incontext_data:
+                
+                # Get example prompt
+                character_info = outlines_prompts_utils.get_relevant_structured_character_info(example['state'].characters)
+                icl_scenario_description = scenario_state_description_with_relevant_char_info(example['state'], character_info)
+                icl_choices_with_outcomes = {}
+                for choice in example['choices']:
+                    # TODO: Include outcome prediction for ICL examples?
+                    icl_choices_with_outcomes[choice] = {'predicted_outcome':None}
+                icl_prompt = comparative_kdma_score_prediction_prompt(icl_scenario_description,
+                                                                    icl_choices_with_outcomes,
+                                                                    sys_kdma_name)
+                # Get example response
+                icl_response = {}
+                for action, choice, kdma_value in zip(example['actions'], example['choices'], example["kdma_values"]):
+                    # Only include choice if there is a ground truth KDMA value available
+                    if kdma_value is None:
+                        continue
+                    # Predicted scores are 0-10, KDMA values are 0-1
+                    scaled_kdma_value = int(kdma_value * 10)
+                    icl_response[choice] = {}
+                    icl_response[choice]['reasoning'] = self.get_chain_of_thought_reasoning(target_kdma, action,
+                                                                                            example['state'], choice,
+                                                                                            scaled_kdma_value)
+                    icl_response[choice]['score'] = scaled_kdma_value
```

Comment:
General comment, not expecting changes: I really wish that we could verify that this is consistent with the json schema, in case that changes

Reply:
Good idea, I added json schema validation here: https://github.com/ITM-Kitware/align-system/pull/115/commits/c5c7c851ff940c4dbee37272a390df8b7b1e000c